File: frameworks/csharp-dotnet/FraiseQL.Benchmark.Tests/TestFactory.cs
using FraiseQL.Benchmark.Models;
using System;
using System.Collections.Generic;
using System.Linq;

namespace FraiseQL.Benchmark.Tests;

public class TestFactory
{
    private readonly Dictionary<Guid, User> _users = new();
    private readonly Dictionary<Guid, Post> _posts = new();
    private int _userCounter = 1;
    private int _postCounter = 1;

    public User CreateTestUser(string username, string email, string fullName, string bio)
    {
        var user = new User
        {
            PkUser = _userCounter++,
            Id = Guid.NewGuid(),
            Username = username,
            FullName = fullName,
            Bio = string.IsNullOrEmpty(bio) ? null : bio,
            CreatedAt = DateTime.UtcNow,
            Posts = new List<Post>(),
            Comments = new List<Comment>()
        };

        _users[user.Id] = user;
        return user;
    }

    public Post CreateTestPost(Guid authorId, string title, string content)
    {
        if (!_users.TryGetValue(authorId, out var author))
        {
            throw new ArgumentException($"Author not found: {authorId}");
        }

        var post = new Post
        {
            PkPost = _postCounter++,
            Id = Guid.NewGuid(),
            Title = title,
            Content = string.IsNullOrEmpty(content) ? string.Empty : content,
            FkAuthor = author.PkUser,
            Author = author,
            CreatedAt = DateTime.UtcNow,
            Comments = new List<Comment>()
        };

        _posts[post.Id] = post;
        return post;
    }

    public Comment CreateTestComment(Guid authorId, Guid postId, string content)
    {
        if (!_users.TryGetValue(authorId, out var author))
        {
            throw new ArgumentException($"Author not found: {authorId}");
        }

        if (!_posts.TryGetValue(postId, out var post))
        {
            throw new ArgumentException($"Post not found: {postId}");
        }

        var comment = new Comment
        {
            Id = Guid.NewGuid(),
            Content = content,
            Author = author,
            Post = post,
            CreatedAt = DateTime.UtcNow
        };

        return comment;
    }

    public User? GetUser(Guid id)
    {
        _users.TryGetValue(id, out var user);
        return user;
    }

    public Post? GetPost(Guid id)
    {
        _posts.TryGetValue(id, out var post);
        return post;
    }

    public IEnumerable<User> GetAllUsers() => _users.Values;

    public IEnumerable<Post> GetAllPosts() => _posts.Values;

    public int GetUserCount() => _users.Count;

    public int GetPostCount() => _posts.Count;

    public void Reset()
    {
        _users.Clear();
        _posts.Clear();
        _userCounter = 1;
        _postCounter = 1;
    }
}

File: frameworks/express-orm/tests/test-factory.ts
import { v4 as uuidv4 } from 'uuid';

export interface TestUser {
  id: string;
  pk_user: number;
  username: string;
  full_name: string;
  bio: string | null;
  created_at: Date;
  updated_at: Date;
}

export interface TestPost {
  id: string;
  pk_post: number;
  fk_author: number;
  title: string;
  content: string;
  published: boolean;
  created_at: Date;
  updated_at: Date;
  author?: TestUser;
}

export interface TestComment {
  id: string;
  pk_comment: number;
  fk_post: number;
  fk_author: number;
  content: string;
  created_at: Date;
  author?: TestUser;
  post?: TestPost;
}

export class TestFactory {
  private users: Map<string, TestUser> = new Map();
  private posts: Map<string, TestPost> = new Map();
  private comments: Map<string, TestComment> = new Map();
  private userCounter: number = 1;
  private postCounter: number = 1;
  private commentCounter: number = 1;

  createTestUser(username: string, email: string, fullName: string, bio: string): TestUser {
    const user: TestUser = {
      id: uuidv4(),
      pk_user: this.userCounter++,
      username,
      full_name: fullName,
      bio: bio ? bio : null,
      created_at: new Date(),
      updated_at: new Date(),
    };

    this.users.set(user.id, user);
    return user;
  }

  createTestPost(authorId: string, title: string, content: string): TestPost {
    const author = this.users.get(authorId);
    if (!author) {
      throw new Error(`Author not found: ${authorId}`);
    }

    const post: TestPost = {
      id: uuidv4(),
      pk_post: this.postCounter++,
      fk_author: author.pk_user,
      title,
      content,
      published: true,
      created_at: new Date(),
      updated_at: new Date(),
      author,
    };

    this.posts.set(post.id, post);
    return post;
  }

  createTestComment(authorId: string, postId: string, content: string): TestComment {
    const author = this.users.get(authorId);
    const post = this.posts.get(postId);

    if (!author || !post) {
      throw new Error('Author or Post not found');
    }

    const comment: TestComment = {
      id: uuidv4(),
      pk_comment: this.commentCounter++,
      fk_post: post.pk_post,
      fk_author: author.pk_user,
      content,
      created_at: new Date(),
      author,
      post,
    };

    this.comments.set(comment.id, comment);
    return comment;
  }

  getUser(id: string): TestUser | undefined {
    return this.users.get(id);
  }

  getPost(id: string): TestPost | undefined {
    return this.posts.get(id);
  }

  getComment(id: string): TestComment | undefined {
    return this.comments.get(id);
  }

  getAllUsers(): TestUser[] {
    return Array.from(this.users.values());
  }

  getAllPosts(): TestPost[] {
    return Array.from(this.posts.values());
  }

  getAllComments(): TestComment[] {
    return Array.from(this.comments.values());
  }

  getPostsByAuthor(authorPk: number): TestPost[] {
    return Array.from(this.posts.values()).filter(p => p.fk_author === authorPk);
  }

  getCommentsByPost(postPk: number): TestComment[] {
    return Array.from(this.comments.values()).filter(c => c.fk_post === postPk);
  }

  getUserCount(): number {
    return this.users.size;
  }

  getPostCount(): number {
    return this.posts.size;
  }

  getCommentCount(): number {
    return this.comments.size;
  }

  reset(): void {
    this.users.clear();
    this.posts.clear();
    this.comments.clear();
    this.userCounter = 1;
    this.postCounter = 1;
    this.commentCounter = 1;
  }
}

export class ValidationHelper {
  static isValidUUID(value: string): boolean {
    const uuidRegex = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;
    return uuidRegex.test(value);
  }

  static assertUUID(value: string): void {
    if (!this.isValidUUID(value)) {
      throw new Error(`Invalid UUID: ${value}`);
    }
  }

  static assertNotEmpty(value: string, name: string): void {
    if (!value || value.trim() === '') {
      throw new Error(`${name} should not be empty`);
    }
  }
}

export class DataGenerator {
  static generateLongString(length: number): string {
    let result = '';
    for (let i = 0; i < length; i++) {
      result += String(i % 10);
    }
    return result;
  }

  static generateUniqueStrings(count: number): string[] {
    return Array.from({ length: count }, () => uuidv4());
  }
}
